.github/workflows/elm_tests.yml

```diff
@@ -8,37 +8,68 @@ jobs:
+      # review/elm.json changes. The Elm compiler saves downloaded Elm packages
+      # to ~/.elm, and elm-tooling saves downloaded tool executables there.
+      - name: Cache ~/.elm
+        uses: actions/cache@9b0c1fce7a93df8e3bb8926b0d6e9d89e92f20a7
         with:
-          elm-version: 0.19.1
+          path: ~/.elm
+          key: elm-${{ hashFiles('elm.json', 'elm-tooling.json', 'review/elm.json') }}
 
-      - name: Install elm-test, elm-format and elm-review
-        run: |
-          npm config set prefix $HOME/.npm-global
-          echo "$HOME/.npm-global/bin" >> $GITHUB_PATH
-          npm install -g elm-test@0.19.1-revision9 elm-format elm-review
+      # Install npm packages, unless we restored them from cache.
+      # Since `npm ci` removes the node_modules folder before running it’s
+      # important to skip this step if cache was restored.
+      # `npm ci` does two things:
+      # 1. Installs everything in package-lock.json.
+      # 2. Checks that package.json and package-lock.json are in sync.
+      # That’s why the cache depends on both package-lock.json and package.json.
+      - name: npm ci
+        if: steps.cache-node_modules.outputs.cache-hit != 'true'
+        env:
+          # If you have a `"postinstall": "elm-tooling install"` script in your
+          # package.json, this turns it into a no-op. We’ll run it in the next
+          # step because of the caching. If elm-tooling.json changes but
+          # package-lock.json does not, the postinstall script needs running
+          # but this step won’t.
+          NO_ELM_TOOLING_INSTALL: 1
+        run: npm ci
+
+      # Install tools from elm-tooling.json, unless we restored them from
+      # cache. package-lock.json and elm-tooling.json can change independently,
+      # so we need to install separately based on what was restored from cache.
+      # This is run even if we restored ~/.elm from cache to be 100% sure
+      # node_modules/.bin/ contains links to all your tools. `elm-tooling
+      # install` runs very fast when there’s nothing new to download so
+      # skipping the step doesn’t save much time.
+      - name: elm-tooling install
+        run: npx elm-tooling install
 
       - name: Run Tests
-        run: elm-test
+        run: npx elm-test-rs
 
       - name: Check formatting
-        run: elm-format --validate src/ tests/ review/
+        run: npx elm-format --validate src/ tests/ review/
 
       - name: Run elm-review on source code
-        run: elm-review
-
-      - name: Run Tests in Docker
-        run: bin/run-tests-in-docker.sh
+        run: npx elm-review
 
   smoke-test:
     runs-on: ubuntu-latest
 
     steps:
       - name: Checkout code
-        uses: actions/checkout@5a4ac9002d0be2fb38bd78e4b4dbde5606d7042f
+        uses: actions/checkout@e2f20e631ae6d7dd3b768f56a5d2af784dd54791
 
       - name: Run Smoke Tests in Docker
         run: bin/run-tests-in-docker.sh
```

Comment:
@ErikSchierboom is there an easy way to cache the docker image? I realized that a new one is built each time.
It only takes a minute to run, but it would still help.

Reply:
I think you can use the docker build push action to cache it, and a flag saying you don't want to push it anywhere.

https://github.com/docker/build-push-action/blob/master/docs/advanced/cache.md

Reply:
Yep, that would be the recommended approach.

Reply:
That looks a bit complicated :p 
Plus, the image does have to rebuild every time we change the source code, and the common parts are already cached. So... I'll give up on that for now :)
I appreciate the advice though.

@ErikSchierboom can I ask you to approve this PR?